File: sdks/swift/KernelyraClient.swift
import Foundation

public final class KernelyraClient {
    private let process = Process()
    private let input = Pipe()
    private let output = Pipe()
    private var requestID: UInt64 = 0
    private let lock = NSLock()

    public init(workspace: String, executable: String = "/usr/bin/env") throws {
        process.executableURL = URL(fileURLWithPath: executable)
        process.arguments = executable == "/usr/bin/env"
            ? ["kernelyra", "--workspace", workspace, "rpc"]
            : ["--workspace", workspace, "rpc"]
        process.standardInput = input
        process.standardOutput = output
        process.standardError = FileHandle.standardError
        try process.run()
        let ready = try readObject()
        guard ready["protocol"] as? String == "kernelyra-jsonl/1" else {
            throw NSError(domain: "Kernelyra", code: 1, userInfo: [NSLocalizedDescriptionKey: "Incompatible protocol"])
        }
    }

    public func call(_ method: String, params: [String: Any] = [:]) throws -> Any {
        lock.lock(); defer { lock.unlock() }
        requestID += 1
        let request: [String: Any] = ["id": requestID, "method": method, "params": params]
        var data = try JSONSerialization.data(withJSONObject: request)
        data.append(0x0A)
        try input.fileHandleForWriting.write(contentsOf: data)
        let response = try readObject()
        guard response["ok"] as? Bool == true else {
            throw NSError(domain: "Kernelyra", code: 2, userInfo: [NSLocalizedDescriptionKey: response["error"] as? String ?? "Kernelyra error"])
        }
        return response["result"] as Any
    }

    public func plan(_ dataset: String, options: [String: Any] = [:]) throws -> Any {
        try call("plan", params: options.merging(["dataset": dataset]) { _, new in new })
    }
    public func train(_ dataset: String, options: [String: Any] = [:]) throws -> Any {
        try call("train", params: options.merging(["dataset": dataset]) { _, new in new })
    }
    public func finetune(_ model: String, dataset: String, options: [String: Any] = [:]) throws -> Any {
        try call("finetune", params: options.merging(["model": model, "dataset": dataset]) { _, new in new })
    }

    private func readObject() throws -> [String: Any] {
        var data = Data()
        while true {
            let byte = output.fileHandleForReading.readData(ofLength: 1)
            if byte.isEmpty { throw NSError(domain: "Kernelyra", code: 3, userInfo: [NSLocalizedDescriptionKey: "Protocol closed"])}
            if byte[0] == 0x0A { break }
            data.append(byte)
        }
        return try JSONSerialization.jsonObject(with: data) as! [String: Any]
    }

    public func close() {
        try? input.fileHandleForWriting.close()
        process.waitUntilExit()
    }

    deinit { if process.isRunning { process.terminate() } }
}

File: sdks/php/KernelyraClient.php
<?php
declare(strict_types=1);

final class KernelyraClient {
    private $process;
    private array $pipes;
    private int $id = 0;

    public function __construct(string $workspace, string $executable = 'kernelyra') {
        $command = [$executable, '--workspace', $workspace, 'rpc'];
        $this->process = proc_open($command, [['pipe','r'], ['pipe','w'], ['pipe','w']], $this->pipes);
        if (!is_resource($this->process)) throw new RuntimeException('Cannot start Kernelyra');
        $ready = json_decode((string) fgets($this->pipes[1]), true, flags: JSON_THROW_ON_ERROR);
        if (($ready['protocol'] ?? '') !== 'kernelyra-jsonl/1') throw new RuntimeException('Incompatible protocol');
    }

    public function call(string $method, array $params = []): array {
        $request = ['id' => ++$this->id, 'method' => $method, 'params' => $params];
        fwrite($this->pipes[0], json_encode($request, JSON_THROW_ON_ERROR) . "\n"); fflush($this->pipes[0]);
        $response = json_decode((string) fgets($this->pipes[1]), true, flags: JSON_THROW_ON_ERROR);
        if (!($response['ok'] ?? false)) throw new RuntimeException((string) ($response['error'] ?? 'Kernelyra error'));
        return $response['result'];
    }

    public function plan(string $dataset, array $options = []): array { return $this->call('plan', ['dataset'=>$dataset] + $options); }
    public function train(string $dataset, array $options = []): array { return $this->call('train', ['dataset'=>$dataset] + $options); }
    public function finetune(string $model, string $dataset, array $options = []): array { return $this->call('finetune', ['model'=>$model, 'dataset'=>$dataset] + $options); }
    public function close(): void { fclose($this->pipes[0]); fclose($this->pipes[1]); fclose($this->pipes[2]); proc_close($this->process); }
}
